### ISSUE TITLE
Icons and labels in a11y

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#12284]
Figure out how Icons should be labeled in a11y mode. Currently, an Icon doesn't have a label and an IconButton for example will not announce itself properly to assistive technologies. There seem to be two ways to fix this:

1. Each Icon should have a default label that it presents to assistive technologies. There needs to be a way to easily override this default label if necessary. If a user wants to exclude an Icon from accessibility, the `ExcludeSemantics` widget can be used.

2. Require users to add the label with a `Semantics` widget.

I am leaning towards option 1, as this will guarantee that IconButtons have a somewhat reasonable label even if the developer didn't think about accessibility.

**Comment by Hixie** [comment]
Option 1 is going to involve a ridiculous amount of work for translators.

**Comment by goderbauer** [comment]
So, it's a trade-off: Either we do more (translation) work in the framework or developers will have to do more work to make their apps truly accessible...

I wonder how native Android/iOS apps deal with Material Icons. Do they provide a default label there? If so, do they provide translations? That might be something worth checking when tackling this issue.

**Comment by Hixie** [comment]
Most apps will have to come up with their own translation strings anyway, since they'll need to write custom tooltips for each button (there's no way we can guess what they're using the icons for).

**Comment by goderbauer** [comment]
Yeah... I18N is going to be difficult. 

Under the aspect of a11y it would be nice if `tooltip` were a required argument on `IconButton`  to force people to provide a textual description of that button, which will be read out by TalkBack/VoiceOver. However, it's probably too late to make that breaking API change?

In practice, that will probably mean that people will only detect missing labels when running their app through accessibility checkers. Android's Accessibility Scanner for example correctly identifies IconButtons with missing labels in Flutter apps.

To summarize: I believe we can close this issue as there is nothing actionable on the Flutter Framework side. The solution is: App developers have to provide proper tooltips for IconButtons and we can help them find missing labels (e.g. today by pointing them toward Android's Accessibility Scanner and in the future potentially by providing our own Flutter Accessibility Inspector).

**Comment by Hixie** [comment]
If you require tooltip, people will just say `tooltip: ''`.

**Comment by Hixie** [comment]
(or worse, things like `tooltip: 'add'`, copied and pasted all over the place)

**Comment by goderbauer** [comment]
Alright. Closing this as there is nothing actionable for Flutter as described in my previous post.

**Comment by eseidelGoogle** [comment]
Presumably we could create (or use/recommend) some ax validation tool which would warn about missing tooltip: or the like.  Lints might also work for such, donno.
